src/Analyzers/CSharp/CodeFixes/UseUtf8StringLiteral/UseUtf8StringLiteralCodeFixProvider.cs

```diff
@@ -48,12 +48,26 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
             var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
 
+            var readOnlySpanType = semanticModel.Compilation.GetBestTypeByMetadataName(typeof(ReadOnlySpan<>).FullName!);
+            // The analyzer wouldn't raise a diagnostic if this were null
+            Contract.ThrowIfNull(readOnlySpanType);
+
             foreach (var diagnostic in diagnostics)
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
                 var node = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
-                var stringValue = GetUtf8StringValueForDiagnostic(semanticModel, diagnostic, cancellationToken);
+                var arrayOp = GetArrayCreationOperation(semanticModel, diagnostic, cancellationToken);
+                Contract.ThrowIfNull(arrayOp.Initializer);
+
+                var stringValue = GetUtf8StringValueFromArrayInitializer(arrayOp.Initializer);
+
+                // If our array is parented by a conversion to ReadOnlySpan<byte> then we don't want to call
+                // ToArray after the string literal, or we'll be regressing perf.
+                var isConvertedToReadOnlySpan = arrayOp.Parent is IConversionOperation conversion &&
+                    conversion.Type is INamedTypeSymbol { IsGenericType: true } namedType &&
+                    namedType.ConstructedFrom == readOnlySpanType &&
```

Comment:
```suggestion
                    namedType.ConstructedFrom.Equals(readOnlySpanType) &&
```

symbols are not guaranteed to have reference-equal equality from the compiler layer.

Reply:
> Can you add a test taht common C# characters (like`\r\n\t`) still work?

Actually, they don't with this change, but probably a good idea to special case them. Will add that